csi/deployment_util.go

```diff
@@ -10,8 +11,142 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/pointer"
 )
 
+const (
+	maxRetryCountForMountPropagationCheck = 10
+	durationSleepForMountPropagationCheck = 10 * time.Second
+)
+
+// CheckMountPropagationWithPodSpec https://github.com/kubernetes/kubernetes/issues/66086#issuecomment-404346854
+func CheckMountPropagationWithPodSpec(kubeClient *clientset.Clientset, image, namespace string) error {
+	commonName := "longhorn-mount-propagation-tester"
+	mountName := "mountpoint"
+	mountPath := "/mnt/tmp"
+	podSpec := v1.PodSpec{
+		Containers: []v1.Container{
+			v1.Container{
+				Name:  commonName,
+				Image: image,
+				Args: []string{
+					"/bin/sh",
+					"-c",
+					"sleep infinity",
+				},
+				ImagePullPolicy: v1.PullIfNotPresent,
+				VolumeMounts: []v1.VolumeMount{
+					v1.VolumeMount{
+						Name:             mountName,
+						MountPath:        mountPath,
+						MountPropagation: &MountPropagationBidirectional,
+					},
+				},
+				SecurityContext: &v1.SecurityContext{
+					Privileged: pointer.BoolPtr(true),
+				},
+			},
+		},
+		Volumes: []v1.Volume{
+			v1.Volume{
+				Name: mountName,
+				VolumeSource: v1.VolumeSource{
+					HostPath: &v1.HostPathVolumeSource{
+						Path: mountPath,
+					},
+				},
+			},
+		},
+	}
+
+	daemonSet := &appsv1beta2.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      commonName,
+			Namespace: namespace,
+		},
+
+		Spec: appsv1beta2.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": commonName,
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": commonName,
+					},
+				},
+				Spec: podSpec,
+			},
+		},
+	}
+
+	daemonSetClient := kubeClient.AppsV1beta2().DaemonSets(namespace)
+	podClient := kubeClient.CoreV1().Pods(namespace)
+
+	logrus.Debugf("Trying to create the daemonset %s for MountPropagation check", commonName)
+	_, err := daemonSetClient.Create(daemonSet)
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("Created the daemonset %s for MountPropagation check", commonName)
+
+	defer func() {
+		propagation := metav1.DeletePropagationForeground
+		if err := daemonSetClient.Delete(commonName,
+			&metav1.DeleteOptions{PropagationPolicy: &propagation}); err != nil {
+			logrus.Warnf("Failed to delete the daemonset %s for MountPropagation check", commonName)
+		}
+	}()
+
+	retryCount := 0
+	for {
+		if retryCount >= maxRetryCountForMountPropagationCheck {
+			return fmt.Errorf("Has been retried %d times, but daemonset is still unavailable", retryCount)
+		}
+		time.Sleep(durationSleepForMountPropagationCheck)
+		retryCount++
+		ds, err := daemonSetClient.Get(commonName, metav1.GetOptions{})
+		if err != nil {
```

Comment:
I think the only valid error here is `apierrors.IsNotFound(err)`. Otherwise, it would be a real problem and we probably don't want to retry.

Reply:
We are getting the daemonset not its pods, is it possible to get IsNotFound error?

Reply:
The race condition is due to a gap between the API server receive the request to create the object, and it actually creates the object and makes the client known. The race condition Jacie saw last time is for replica CRD object, so I think it should applies to all the object. Seems creation request is not a blocking command and won't guarantee the immediate request after creation can see the object ready.

Reply:
Done.

But this may be slightly different, Jacie used List, and here is Get.
List reads data from the cache, so sometimes it is not possible to get the data right away.
But if adding this code to every Get, it looks ugly.

I will confirm how to use Get properly.

Reply:
Yeah, that's possible too. But even for the lister, it shouldn't so hard to use as well. I thought maybe it's because of the eventual consistency of the etcd, but not entirely sure about that.

Please let me know what you find in the end.